protocol/l1-upgrades.md

```diff
@@ -0,0 +1,338 @@
+the summary should be as succinct as possible. -->
+
+The proposed solution will perform upgrades by transferring upgrade authorization of all L1
+contracts to the release specific OPCM. Upon completion of the upgrade, this authorization will
+be returned to the Security Council.
+
+Rather than the curent `StorageSetter` based approach which simply deletes the `_initialized` slot,
+allowing for `initialize()` to be called, we take this two-step upgrade further, with a short-lived
+`Populator` contract being temporarily set as the implementation in order to set or modify any
+values as required. This has two significant benefits. First it removes the need to keep a one-time
+`initialize()` function in the run-time logic. Secondly it enables a clean separation of concerns
+between contracts which are newly deployed (requiring all values to be set), and contracts being
+upgraded (requiring only new values to be set).
+
+# Problem Statement + Context
+
+Our upgrades have the following issues:
+
+1. They are divorced from the implementation, meaning the work of writing the upgrade script occurs
+   in `superchain-ops` apart from the work of writing the updated code in the `optimism` monorepo.
+2. The act of writing the upgrade script is onerous, requiring:
+   1. A bespoke script with custom assertions.
+   2. A bespoke Validations.md document
+3. The work required to perform the upgrade is difficult for signers, as they must work through the
+   Validations.md document. This work has taken up to an hour in the past.
+4. The effort described above will scale linearly with the number of OP Chains.
+
+# Proposed Solution
+
+Each new release will have its own OP Contracts Manager which can perform the following distinct
+actions:
+
+1. Deploy new proxies for a new OP Chain (this work is already covered by [link to Blaine's doc]())
+1. Upgrade all existing OP Chains.
+1. Deploy any newly added superchain-shared contracts.
+1. Upgrade existing superchain-shared contracts.
+
+Note that the above functionality also makes for a good set of milestones, each of which can and
+should be implemented incrementally.
+
+## Declarative storage management
+
+Taking inspiration from Chugsplash, we wish to separate the concern of what bytecode to run, and
+what to write to storage. However we also wish to avoid using the current `StorageSetter` pattern,
+which simply provides logic to "write anything anywhere" in storage.
+
+In order to achieve this, each contract will have its `initializer()` function moved intoa
+release specific `__Populator` contract.
+
+By way of example, we take an initializable version of `SimpleStorage`:
+
+- the current version of the contract has only a `uint256 foo` variable.
+- the new version of the contract introduces a `bytes32 bar` variable.
+
+The status quo implementation of the new Simple Storage would look like this:
+
+```solidity
+contract SimpleStorage is Initializable {
```

Comment:
How much more difficult would it be to maintain using `StorageSetter` instead of using the `Populator` pattern? There are 2 paths where this code needs to be used. Once with new chains and once with upgrading chains.

For new chains, it would be set all storage and set the storage version to latest

For upgrading chains, it would be check storage version is 1 less than latest, add/move/remove any slots, then set the storage version to latest

There are 2 chunks of logic which may be different no matter what

Reply:
The 2 chunks of logic are represented in `upgrade` vs `initialize`, this makes sense

Reply:
> How much more difficult would it be to maintain using StorageSetter instead of using the Populator pattern? There are 2 paths where this code needs to be used. Once with new chains and once with upgrading chains.

I believe it would be very difficult to replace [`SystemConig.initialize()`](https://github.com/ethereum-optimism/optimism/blob/maur/repeat-isthmus-contracts/packages/contracts-bedrock/src/L1/SystemConfig.sol#L185-L229) (which accepts 20 values) with the `StorageSetter`. Writing all the values is unergononomic but doable, however the many sanity checks on those values would need to be moved into the OPCM which would be difficult to maintain and reason about.

Reply:
Resolving as we're no long proposing the `Populator` and just adding `upgrade()` to the impl.